types/src/transaction/transaction.rs

```diff
@@ -264,6 +232,115 @@ fn is_input_and_output_consistent(inputs: &[AssetTransferInput], outputs: &[Asse
     sum.iter().all(|(_, sum)| sum.is_zero())
 }
 
+fn make_tag(mut item: Vec<u8>, mut result: u128) -> u128 {
+    if item.is_empty() {
+        return result
+    }
+    // Assume that the mask is stored as little endian
+    let head = item.pop().unwrap();
+    result = (result << 8) | head as u128;
+    make_tag(item, result)
+}
+
+
+fn apply_bitmask_to_output(
+    mut bitmask: u128,
+    outputs: Vec<AssetTransferOutput>,
+    mut result: Vec<AssetTransferOutput>,
+) -> Result<Vec<AssetTransferOutput>, HashingError> {
+    for i in 0.. {
+        if bitmask == 0 {
+            return Ok(result)
+        }
+
+        if outputs.is_empty() {
+            return Err(HashingError::InvalidFilter)
+        }
+
+        if (bitmask & 0x1) == 1 {
+            result.push(outputs[i].clone());
+        }
+        bitmask = bitmask >> 1
+    }
+    Err(HashingError::InvalidFilter)
+}
+
+fn apply_sign_scheme_to_input(
+    inputs: &Vec<AssetTransferInput>,
+    scheme: bool,
+    is_burn: bool,
+    cur: &AssetOutPoint,
+) -> Vec<AssetTransferInput> {
+    if scheme {
+        return inputs
+            .iter()
+            .map(|input| AssetTransferInput {
+                prev_out: input.prev_out.clone(),
+                lock_script: Vec::new(),
+                unlock_script: Vec::new(),
+            })
+            .collect()
+    }
+
+    if is_burn {
+        return vec![AssetTransferInput {
+            prev_out: cur.clone(),
+            lock_script: Vec::new(),
+            unlock_script: Vec::new(),
+        }]
+    }
+
+    Vec::new()
+}
+
+impl PartialHashing for Transaction {
+    fn hash_partially(&self, bitvec: Vec<u8>, cur: &AssetOutPoint, burn: bool) -> Result<H256, HashingError> {
+        let tag = make_tag(bitvec.clone(), 0);
+        let sign_all_inputs = (tag & 0x1) == 1;
+        let sign_all_outputs = (tag >> 1 & 0x1) == 1;
+        let mask = tag >> 2;
+
+        match self {
+            Transaction::AssetTransfer {
+                network_id,
+                burns,
+                inputs,
+                outputs,
+                nonce,
+            } => {
+                let new_burns = apply_sign_scheme_to_input(burns, sign_all_inputs, burn, cur);
+                let new_inputs = apply_sign_scheme_to_input(inputs, sign_all_inputs, burn, cur);
+
+                let new_outputs = if sign_all_outputs {
+                    outputs
+                        .iter()
+                        .map(|output| AssetTransferOutput {
```

Comment:
Why not `.map(Clone::clone)`?

Reply:
It seems that it can be `outputs.clone()` if it can be `.map(Clone::clone).

Reply:
It is my mistake. I didn't recover that from the old scheme. In the previous version I implemented, there was an additional element for storing filter in output and input. So I had to clone other elements one by one like this way. However, there is no need to do this now.